lib/ddtrace/profiling/collectors/stack.rb

```diff
@@ -0,0 +1,111 @@
+require 'ddtrace/profiling/events/stack'
+require 'ddtrace/utils/time'
+require 'ddtrace/worker'
+require 'ddtrace/workers/polling'
+
+module Datadog
+  module Profiling
+    module Collectors
+      # Pulls and exports profiling data on an async interval basis
+      class Stack < Worker
+        include Workers::Polling
+
+        MIN_INTERVAL = 0.01
+        MAX_FRAMES = 128
+
+        attr_reader \
+          :ignore_thread,
+          :last_wall_time,
+          :max_frames,
+          :max_time_usage_pct,
+          :recorder
+
+        def initialize(recorder, options = {})
+          @recorder = recorder
+          @max_frames = options.fetch(:max_frames, MAX_FRAMES)
+          @ignore_thread = options.fetch(:ignore_thread, nil)
+          @max_time_usage_pct = options.fetch(:max_time_usage_pct, 2.0)
+
+          # Workers::Async::Thread settings
+          # Restart in forks by default
+          self.fork_policy = options.fetch(:fork_policy, Workers::Async::Thread::FORK_POLICY_RESTART)
+
+          # Workers::IntervalLoop settings
+          self.loop_base_interval = options.fetch(:interval, MIN_INTERVAL)
+
+          # Workers::Polling settings
+          self.enabled = options.fetch(:enabled, false)
+        end
+
+        def start
+          perform
+        end
+
+        def perform
+          collect_and_wait
+        end
+
+        def loop_back_off?
+          false
+        end
+
+        def collect_and_wait
+          run_time = Datadog::Utils::Time.measure do
+            collect_events
+          end
+
+          # Update wait time to throttle profiling
+          self.loop_wait_time = compute_wait_time(run_time)
+        end
+
+        def collect_events
+          @last_wall_time ||= Datadog::Utils::Time.get_time
+          events = []
+
+          # Compute wall time interval
+          current_wall_time = Datadog::Utils::Time.get_time
+          wall_time_interval_ns = (current_wall_time - last_wall_time) * 1e9
+          @last_wall_time = current_wall_time
+
+          # Collect backtraces from each thread
+          Thread.list.each do |thread|
+            next unless thread.alive?
+            next if ignore_thread.is_a?(Proc) && ignore_thread.call(thread)
+
+            event = collect_thread_event(thread, wall_time_interval_ns)
+            events << event unless event.nil?
+          end
+
+          # Send events to recorder
+          recorder.push(events) unless events.empty?
+
+          events
+        end
+
+        def collect_thread_event(thread, wall_time_interval_ns)
+          locations = thread.backtrace_locations
+          return if locations.nil?
+
+          # Get actual stack size then trim the stack
+          stack_size = locations.length
+          locations = locations[0..max_frames]
+
+          Events::StackSample.new(
+            nil,
+            locations,
+            stack_size,
+            nil,
```

Comment:
This is `nil` because we can't get the native thread ID to get the CPU time. We could probably replace it with the `thread.object_id` instead, as a stand in.